framework/src/service/ServiceListeners.cpp

```diff
@@ -82,82 +93,170 @@ void ServiceListeners::Clear()
+}
+
+ListenerToken ServiceListeners::AddBundleListener(const std::shared_ptr<BundleContextPrivate>& context, const BundleListener& listener, void* data)
+{
+  auto token = MakeListenerToken();
+
+  auto l = bundleListenerMap.Lock(); US_UNUSED(l);
+  auto& listeners = bundleListenerMap.value[context];
+  listeners.push_back(std::make_tuple(listener, data, token.getId()));
+  return token;
 }
 
+/**
+ * Called by the deprecated RemoveBundleListener(name_of_callable)
+ */
 void ServiceListeners::RemoveBundleListener(const std::shared_ptr<BundleContextPrivate>& context, const BundleListener& listener, void* data)
 {
   auto l = bundleListenerMap.Lock(); US_UNUSED(l);
-  bundleListenerMap.value[context].remove_if(std::bind(BundleListenerCompare(), std::make_pair(listener, data), std::placeholders::_1));
+  auto& listeners = bundleListenerMap.value[context];
+  listeners.erase(std::remove_if(listeners.begin(),
+                                 listeners.end(),
+                                 std::bind(BundleListenerCompareName, listener, data, std::placeholders::_1)),
+                  listeners.end());
 }
 
-void ServiceListeners::AddFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
+ListenerToken ServiceListeners::AddFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
 {
+  auto token = MakeListenerToken();
+
   auto l = frameworkListenerMap.Lock(); US_UNUSED(l);
   auto& listeners = frameworkListenerMap.value[context];
-  if (std::find_if(listeners.begin(), listeners.end(), std::bind(FrameworkListenerCompare(), std::make_pair(listener, data), std::placeholders::_1)) == listeners.end())
-  {
-    listeners.push_back(std::make_pair(listener, data));
-  }
+  listeners.push_back(std::make_tuple(listener, data, token.getId()));
+  return token;
 }
 
+/**
+ * Called by the deprecated RemoveFrameworkListener(name_of_callable)
+ */
 void ServiceListeners::RemoveFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
 {
   auto l = frameworkListenerMap.Lock(); US_UNUSED(l);
   auto& listeners = frameworkListenerMap.value[context];
   listeners.erase(std::remove_if(listeners.begin(),
                                  listeners.end(),
-                                 std::bind(FrameworkListenerCompare(), std::make_pair(listener, data), std::placeholders::_1)),
+                                 std::bind(FrameworkListenerCompareName, listener, data, std::placeholders::_1)),
                   listeners.end());
 }
 
+/**
+ * Called by RemoveListener(listenerToken)
+ * Iterates over the set of FrameworkListeners, BundleListeners and ServiceListeners to find the tokenId and
+ * remove the listener corresponding to the tokenId.
+ */
+void ServiceListeners::RemoveListener(const std::shared_ptr<BundleContextPrivate>& context, const ListenerToken& token)
+{
+  {
```

Comment:
Your comment about changing the Bundle and Framework listeners to be an `unordered_map` with the `ListenerTokenId`s as the keys sent me in the right path. With this, the amortized removal when using tokens to remove is `O(1)`. Therefore, I reckon we don't even need the bit flag in the token id - because the information that the token id not present in the list of bundle or framework listeners is got after `O(1)`

I am now figuring out if we can follow similar approach for the ServiceListeners.